File: framework/languages/dotnet/e2e/ResilienceLifecycle/Client/Scenarios/RlA4DrainAndGreenEndpointScenario.cs
// Verifies RL-A4 rolling replacement with a ready green target before old-target drain.
using System.Collections.Concurrent;
using ResilienceLifecycle.Client.Support;
using ResilienceLifecycle.Shared;
using Zlink.HttpClient;

namespace ResilienceLifecycle.Client.Scenarios;

// RL-A4 verifies a serving green target before each old provider drains.
internal static class RlA4DrainAndGreenEndpointScenario
{
    public static async Task RunAsync(
        ZLinkHttpClient consumer,
        ZLinkHttpClient registry,
        ResilienceProcessManager processes,
        ZLinkHttpClient providerA,
        ZLinkHttpClient providerB)
    {
        var connectionCount = (await consumer.Get("/connections").Async<string[]>()).Body.Length;
        var green = await processes.StartProviderBGreenAsync();
        using var greenProvider = ZLinkHttpClient.Create(green.Url).Timeout(TimeSpan.FromMinutes(5)).Build();
        await registry.Post("/topology/wait")
            .Body(new TopologyWaitReq("api-green", "Ready", 1))
            .Async<TopologyEntryRes[]>();
        await consumer.Post("/connections/wait")
            .Body(new ConnectionWaitReq(
                ["kind=ConnectionReady", $"remote={green.Endpoint}"], connectionCount))
            .Async<string[]>();
        await ProviderTrafficProbe.DriveUntilProviderServesAsync(
            consumer, greenProvider, "rl-a4-green-ready", "RL-A4 green readiness");

        var replies = new ConcurrentQueue<ProfileRes>();
        var failures = new ConcurrentQueue<Exception>();
        using var trafficStop = new CancellationTokenSource();
        var traffic = DriveTrafficAsync(consumer, replies, failures, trafficStop.Token);

        await DrainAndStopAsync(providerB, registry, "api-b", processes.WaitInitialProviderBExitedAsync);
        await AssertServingAsync(registry, "api-green");
        await DrainAndStopAsync(providerA, registry, "api-a", processes.WaitInitialProviderAExitedAsync);
        await AssertServingAsync(registry, "api-green");

        trafficStop.Cancel();
        await traffic;
        ZlinkStreamAssert.Ensure(failures.IsEmpty,
            $"RL-A4 continuous traffic failed: {failures.FirstOrDefault()?.Message}");
        ZlinkStreamAssert.Ensure(replies.Count >= 10, "RL-A4 did not observe enough fixed-interval traffic.");

        for (var index = 0; index < 20; index++)
        {
            var reply = (await consumer.Post("/profile/request")
                .Body(new ProfileReq("fast", $"rl-a4-final-{index}"))
                .Async<ProfileRes>()).Body;
            ZlinkStreamAssert.Ensure(reply.ProviderRid == "api-green",
                "RL-A4 final traffic was handled by an old provider.");
        }
        await greenProvider.Post("/evidence/wait")
            .Body(new EvidenceWaitReq(["profile-request|rid=api-green|marker=rl-a4-final-19"], []))
            .Async<string[]>();

        Console.WriteLine("scenario RL-A4 passed");
    }

    private static async Task DrainAndStopAsync(
        ZLinkHttpClient provider,
        ZLinkHttpClient registry,
        string rid,
        Func<Task> waitExited)
    {
        var drainTask = provider.Post("/admin/graceful-drain").Async<DrainResultRes>().AsTask();
        await registry.Post("/topology/wait")
            .Body(new TopologyWaitReq(rid, "Ready", 1, ExpectedDraining: true))
            .Async<TopologyEntryRes[]>();
        var result = (await drainTask).Body;
        ZlinkStreamAssert.Ensure(result.Result == "Drained",
            $"RL-A4 {rid} did not reach terminal Drained: {result.Result}/{result.Reason}.");
        await waitExited();
        await registry.Post("/topology/wait")
            .Body(new TopologyWaitReq(rid, "Ready", 0))
            .Async<TopologyEntryRes[]>();
    }

    private static async Task AssertServingAsync(ZLinkHttpClient registry, string rid)
    {
        var rows = (await registry.Post("/topology/wait")
            .Body(new TopologyWaitReq(rid, "Ready", 1, ExpectedDraining: false))
            .Async<TopologyEntryRes[]>()).Body;
        ZlinkStreamAssert.Ensure(rows.Length == 1, "RL-A4 lost its serving green target.");
    }

    private static async Task DriveTrafficAsync(
        ZLinkHttpClient consumer,
        ConcurrentQueue<ProfileRes> replies,
        ConcurrentQueue<Exception> failures,
        CancellationToken cancellationToken)
    {
        var index = 0;
        while (!cancellationToken.IsCancellationRequested)
        {
            try
            {
                var reply = (await consumer.Post("/profile/request")
                    .Body(new ProfileReq("fast", $"rl-a4-continuous-{index++}"))
                    .Async<ProfileRes>(cancellationToken)).Body;
                replies.Enqueue(reply);
            }
            catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
            {
                break;
            }
            catch (Exception error)
            {
                failures.Enqueue(error);
            }

            try
            {
                await Task.Delay(TimeSpan.FromMilliseconds(50), cancellationToken);
            }
            catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
            {
                break;
            }
        }
    }
}

File: framework/languages/node/e2e/ResilienceLifecycle/Client/Scenarios/rl-a4-drain-and-green-endpoint-scenario.ts
// RL-A4: Rolling과 blue-green update 중 serving target을 유지한다 시나리오를 검증한다.
import type { ProfileRes } from '../../Shared/messages';
import type { ClientOptions } from '../Support/client-options';
import { getJson, postJson } from '../../../http-client';
import { startProvider } from '../Support/managed-provider';
import {
  profileReq,
  waitPeerEndpointAbsent,
  waitPeerEndpointPresent,
  waitPeerPresent,
  waitUntilAvailable,
  waitUntilDown
} from '../Support/resilience-helpers';
import { ensure } from '../Support/scenario-assert';
import type { ScenarioState } from '../Support/scenario-state';

export async function runRlA4(options: ClientOptions, state: ScenarioState): Promise<void> {
  await postJson(options.providerBUrl, '/admin/drain');
  await waitForWeight(options.providerBUrl, 0);

  const green = startProvider({
    providerMain: options.providerMain,
    logDir: options.logDir,
    redisEndpoint: options.redisEndpoint,
    redisKeyPrefix: options.redisKeyPrefix,
    name: 'api-b-green',
    rid: 'api-b',
    httpUrl: options.providerBGreenUrl,
    channelEndpoint: options.providerBGreenChannelEndpoint,
    evidenceFileName: 'api-b-green.evidence.log'
  });

  try {
    await green.waitReady();

    for (let i = 0; i < 12; i += 1) {
      const reply = await postJson<ProfileRes>(
        options.consumerUrl,
        '/profile/request',
        profileReq(`rl-a4-rolling-${i}`)
      );
      ensure(reply.value === 'profile:fast', 'RL-A4 rolling request returned an unexpected value.');
      ensure(reply.providerRid === 'api-a' || reply.providerRid === 'api-b', 'RL-A4 rolling request used an unexpected provider.');
    }

    if (state.providerBProcess !== undefined) {
      await state.providerBProcess.stop();
    } else {
      await postJson(options.providerBUrl, '/shutdown');
    }
    await waitUntilDown(options.providerBUrl);
    state.providerBProcess = undefined;
    await waitPeerEndpointAbsent(options.peerLocationUrl, options.providerBChannelEndpoint);

    await waitPeerEndpointPresent(options.peerLocationUrl, 'api-b', options.providerBGreenChannelEndpoint);
    await waitPeerPresent(options.peerLocationUrl, 'api-b');
    for (let i = 0; i < 32; i += 1) {
      const reply = await postJson<ProfileRes>(
        options.consumerUrl,
        '/profile/request',
        profileReq(`rl-a4-green-${i}`)
      );
      ensure(reply.value === 'profile:fast', 'RL-A4 green request returned an unexpected value.');
    }
    await waitEvidenceContains(
      options.providerBGreenUrl,
      'profile-request|rid=api-b|value=fast|marker=rl-a4-green-',
      'RL-A4 green provider did not receive green traffic.'
    );
  } finally {
    await green.stop();
  }
  await waitUntilDown(options.providerBGreenUrl);
  await waitPeerEndpointAbsent(options.peerLocationUrl, options.providerBGreenChannelEndpoint);

  const restored = startProvider({
    providerMain: options.providerMain,
    logDir: options.logDir,
    redisEndpoint: options.redisEndpoint,
    redisKeyPrefix: options.redisKeyPrefix,
    name: 'api-b-original-restored',
    rid: 'api-b',
    httpUrl: options.providerBUrl,
    channelEndpoint: options.providerBChannelEndpoint,
    evidenceFileName: 'api-b-original-restored.evidence.log'
  });
  await restored.waitReady();
  state.providerBProcess = restored;
  await waitUntilAvailable(options.providerBUrl);
  await waitPeerEndpointPresent(options.peerLocationUrl, 'api-b', options.providerBChannelEndpoint);
  await waitPeerPresent(options.peerLocationUrl, 'api-b');

  await sendUntilProvider(options.consumerUrl, 'rl-a4-restored', 'api-b', 120);

  console.log('scenario RL-A4 passed');
}

async function waitForWeight(providerUrl: string, expected: number): Promise<void> {
  await postJson(providerUrl, '/admin/weight/wait', { expected });
}

async function sendUntilProvider(
  consumerUrl: string,
  markerPrefix: string,
  expectedProviderRid: string,
  maxAttempts: number
): Promise<void> {
  for (let i = 0; i < maxAttempts; i += 1) {
    const reply = await postJson<ProfileRes>(
      consumerUrl,
      '/profile/request',
      profileReq(`${markerPrefix}-${i}`)
    );
    ensure(reply.value === 'profile:fast', `${markerPrefix} request returned an unexpected value.`);
    if (reply.providerRid === expectedProviderRid) {
      return;
    }
    await new Promise((resolve) => setTimeout(resolve, 100));
  }
  throw new Error(`Timed out waiting for ${markerPrefix} traffic to reach ${expectedProviderRid}.`);
}

async function waitEvidenceContains(providerUrl: string, contains: string, message: string): Promise<void> {
  const deadline = Date.now() + 10000;
  while (Date.now() < deadline) {
    const evidence = await getJson<string[]>(providerUrl, '/evidence');
    if (evidence.some((line) => line.includes(contains))) {
      return;
    }
    await new Promise((resolve) => setTimeout(resolve, 100));
  }
  throw new Error(message);
}
